sky/resources.py

```diff
@@ -168,6 +169,10 @@ def __init__(
         self._is_image_managed = _is_image_managed
 
         self._disk_tier = disk_tier
+        if ports is not None:
+            if not isinstance(ports, list):
+                ports = [ports]
+            ports = [str(port) for port in ports]
```

Comment:
Why don't we directly save the simplified ports in resources?

Reply:
Because it needs to be aligned with provider configs (to keep track of every rule we created)...
Provider config is where the cleanup refers to, it will iterate every entry of the `ports` section in the provider config, and every one of them is a separate firewall rule that needs to be cleaned up. Simultaneously, the provider config is written according to the `to_provision`, whose ports section cannot be merged.

For example, if we have 10022 and 10024 opened previously, it will become two rules named `user-ports-cluster-10022` and `user-ports-cluster-10024`; if we opened 10023 now then there will be 3 rules (`user-ports-cluster-10022`, `user-ports-cluster-10023`, `user-ports-cluster-10024`) in the cloud, rather than a single rule (`user-ports-cluster-10022-10024`). We cannot track it if we merge it in `to_provision`.